tensorflow/lite/micro/kernels/xtensa_hifi/conv.cc

```diff
@@ -0,0 +1,561 @@
+    output_data = GetTensorData<uint8_t>(output);
+
+    const int stride_width = params->stride_width;
+    const int stride_height = params->stride_height;
+    const int dilation_width_factor = 1;
+    const int dilation_height_factor = 1;
+    const int pad_width = data->padding.width;
+    const int pad_height = data->padding.height;
+    const int32 output_activation_min = data->output_activation_min;
+    const int32 output_activation_max = data->output_activation_max;
+    const int32 output_multiplier = data->output_multiplier;
+    const int output_shift = -data->output_shift;
+    TFLITE_DCHECK_EQ(input_shape.DimensionsCount(), 4);
+    TFLITE_DCHECK_EQ(filter_shape.DimensionsCount(), 4);
+    TFLITE_DCHECK_EQ(output_shape.DimensionsCount(), 4);
+    TFLITE_DCHECK_LE(output_activation_min, output_activation_max);
+
+    const int batches      = MatchingDim(input_shape, 0, output_shape, 0);
+    const int input_depth  = MatchingDim(input_shape, 3, filter_shape, 3);
+    const int output_depth = MatchingDim(filter_shape, 0, output_shape, 3);
+    if (bias_data) {
+      TFLITE_DCHECK_EQ(bias_shape.FlatSize(), output_depth);
+    }
+    const int input_height  = input_shape.Dims(1);
+    const int input_width   = input_shape.Dims(2);
+    const int filter_height = filter_shape.Dims(1);
+    const int filter_width  = filter_shape.Dims(2);
+    const int output_height = output_shape.Dims(1);
+    const int output_width  = output_shape.Dims(2);
+    const int filter_depth  = filter_shape.Dims(3);
+
+    int err, output_data_format=0;
+    void *p_scratch;
+    uint8 *p_filter, *p_out_scratch;
+    // Calculate filter_depth_padded as next near multiple of 4
+    int filter_depth_padded = (filter_depth + 3) & (~3);
+    int out_length = output_height * output_width * output_depth;
+    int required_scratch, input_precision = PREC_ASYM8;
+    int h,w,c;
+
+    required_scratch =  xa_nn_conv2d_std_getsize(
+        input_height,
+        input_depth,
+        filter_height,
+        filter_width,
+        stride_height,
+        pad_height,
+        output_height,
+        input_precision);
+
+    if(required_scratch <= 0){
+      context->ReportError(
+          context, "conv2d_std_asym8: xa_nn_conv2d_std_getsize failed");
+      return kTfLiteError;
+    }
+
+    ALLOCATE_XTENSA_NNLIB_SCRATCH_MEM;
+    p_scratch = xtensa_nnlib_scratch_buf;
+
+    p_filter = (uint8 *)p_scratch;
+    p_out_scratch = (p_filter + ALIGNED_SIZE((sizeof(uint8_t)*filter_height*filter_width*filter_depth_padded*output_depth), 8));
+    required_scratch += ALIGNED_SIZE((sizeof(uint8_t)*filter_height*filter_width*filter_depth_padded*output_depth), 8);
+    p_scratch = (uint8 *)(p_out_scratch + ALIGNED_SIZE(sizeof(uint8_t)*out_length, 8));
+    required_scratch += ALIGNED_SIZE(sizeof(uint8_t)*out_length, 8);
+
+    if(required_scratch > (int)XTENSA_NNLIB_MAX_SCRATCH_SIZE){
+      context->ReportError(context, "conv2d_std_asym8: insufficient scratch memory");
```

Comment:
`TF_LITE_KERNEL_LOG should be used instead of context->ReportError, so log strings can be removed to save space if needed.`